internal/api/kubernetes/client.go

```diff
@@ -510,93 +522,174 @@ func gvrAndKeyFromObj(
-		// internalClient.
-		return internalClient, nil
-	}
+	return func(
+		ctx context.Context,
+		internalClient libClient.Client,
+		verb string,
+		gvr schema.GroupVersionResource,
+		subresource string,
+		key libClient.ObjectKey,
+	) (libClient.Client, error) {
+		userInfo, ok := user.InfoFromContext(ctx)
+		if !ok {
+			return nil, errors.New("not allowed")
+		}
 
-	// If we get to here, we're dealing with a user who "authenticated" by just
-	// passing their bearer token for the Kubernetes API server.
+		// Admins get to use the Kargo API server's own Kubernetes client. i.e. They
+		// can do everything the server can do.
+		if userInfo.IsAdmin {
+			return internalClient, nil
+		}
 
-	// Calling getUserClient will build a user-specific client and conduct a
-	// SelfSubjectAccessReview. If the review succeeds, we'll have obtained a
-	// client that can perform the desired operation. If the review fails, the
-	// returned client will be nil.
-	client, err := getUserClient(
-		ctx,
-		internalClient.Scheme(),
-		userInfo.BearerToken,
-		authv1.ResourceAttributes{
+		ra := authv1.ResourceAttributes{
 			Verb:        verb,
 			Group:       gvr.Group,
 			Version:     gvr.Version,
 			Resource:    gvr.Resource,
 			Subresource: subresource,
 			Namespace:   key.Namespace,
 			Name:        key.Name,
-		},
-	)
-	if err != nil {
-		return nil,
-			errors.Wrap(err, "error getting user-specific Kubernetes client")
+		}
+		if userInfo.Username != "" {
+			serviceAccounts := make([]types.NamespacedName, 0)
+			for _, ns := range []string{
+				key.Namespace,
+				kargoNamespace,
+			} {
+				if ns == "" {
+					continue
+				}
+				accounts, ok := userInfo.ServiceAccounts[ns]
+				if !ok {
+					continue
+				}
+				for _, name := range accounts {
+					serviceAccounts = append(serviceAccounts, types.NamespacedName{
+						Namespace: ns,
+						Name:      name,
+					})
+				}
+			}
+			// If the operation is cluster-scoped, all ServiceAccounts are candidates.
+			if key.Namespace == "" {
+				for ns, accounts := range userInfo.ServiceAccounts {
+					if ns == key.Namespace || ns == kargoNamespace {
+						continue
+					}
+					for _, name := range accounts {
+						serviceAccounts = append(serviceAccounts, types.NamespacedName{
+							Namespace: ns,
+							Name:      name,
+						})
+					}
+				}
+			}
+			for _, sa := range serviceAccounts {
+				_, err := getUserClient(
+					ctx,
+					internalClient.Scheme(),
+					ra,
+					withServiceAccount(sa.Namespace, sa.Name),
+				)
+				if err == nil {
+					// Note: At present, we never return the user-specific client. We always
+					// return the internal client so that most operations are performed by a
+					// single client with a built-in cache.
+					return internalClient, nil
+				}
+				if !apierrors.IsForbidden(err) {
+					return nil, errors.Wrap(err, "get service account specific client")
+				}
+			}
+			return nil, newForbiddenError(ra)
+		}
```

Comment:
Instead of only selecting relevant SAs in the loop above and then doing SARs this additional loop, we could probably have the first loop do the SARs directly.

That would spend less time building a where many SAs may never get checked since the loop short circuits as soon one appropriate SA is found.